umode-api/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 by Rivos Inc.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#![no_std]
+
+//! Salus Hypervisor <-> U-mode API.
+//!
+//! This library contains data structures that are passed between
+//! hypervisor and umode.
+
+/// The Error type returned returned from this library.
+#[derive(Debug, Clone, Copy)]
+#[repr(u64)]
+pub enum Error {
+    /// Generic failure in execution.
+    Failed = 1,
+    /// Ecall not supported. From hypervisor to umode.
+    EcallNotSupported = 2,
+    /// Request not supported. From umode to hypervisor.
+    RequestNotSupported = 3,
+}
+
+impl From<u64> for Error {
+    fn from(val: u64) -> Error {
+        match val {
+            1 => Error::Failed,
+            2 => Error::EcallNotSupported,
+            3 => Error::RequestNotSupported,
+            _ => Error::Failed,
+        }
+    }
+}
+
+// All types that can be passed in registers must implement `IntoRegisters` or `TryIntoRegisters`.
+
+/// Trait to transform a type into A-registers when a set of registers will always transform into
+/// this type.
+pub trait IntoRegisters {
+    /// Get current type from a set of registers.
+    fn from_registers(regs: &[u64]) -> Self;
+    /// Write `self` into a set of registers.
+    fn set_registers(&self, regs: &mut [u64]);
```

Comment:
nit: I'd call this `to_registers` as that sounds like the more natural complement to `from_registers`.

Reply:
Agreed.

Reply:
Done.